pcsx2/USB/usb-pad/usb-mascon.cpp

```diff
@@ -0,0 +1,585 @@
+			// { control_in, emulated_out },
+			{0xF7, 0xFB},
+			{0xE4, 0xE9},
+			{0xD1, 0xD7},
+			{0xBE, 0xC6},
+			{0xAB, 0xB4},
+			{0x98, 0xA2},
+			{0x85, 0x90},
+			{0x72, 0x7E},
+			{0x5F, 0x6C},
+			{0x4C, 0x5A},
+			{0x39, 0x48},
+			{0x26, 0x36},
+			{0x13, 0x24},
+			{0x00, 0x12},
+		};
+
+		for (const auto& x : notches)
+		{
+			if (value >= x.first)
+				return x.second;
+		}
+		return notches[std::size(notches) - 1].second;
+	}
+	static u8 dct02_brake(u8 value)
+	{
+		// (NB) 0x1C 0x38 0x54 0x70 0x8B 0xA7 0xC3 0xDF 0xFB (EB)
+		static std::pair<u8, u8> const notches[] = {
+			// { control_in, emulated_out },
+			{0xF8, 0xFB},
+			{0xCA, 0xDF},
+			{0xAE, 0xC3},
+			{0x92, 0xA7},
+			{0x76, 0x8B},
+			{0x5A, 0x70},
+			{0x3E, 0x54},
+			{0x22, 0x38},
+			{0x00, 0x1C},
+		};
+
+		for (const auto& x : notches)
+		{
+			if (value >= x.first)
+				return x.second;
+		}
+		return notches[std::size(notches) - 1].second;
+	}
+
+	static u8 dct03_power(u8 value)
+	{
+		// (N) 0x00 0x3C 0x78 0xB4 0xF0 (P4)
+		static std::pair<u8, u8> const notches[] = {
+			// { control_in, emulated_out },
+			{0xC0, 0xF0},
+			{0x90, 0xB4},
+			{0x50, 0x78},
+			{0x30, 0x3C},
+			{0x00, 0x00},
+		};
+
+		for (const auto& x : notches)
+		{
+			if (value >= x.first)
+				return x.second;
+		}
+		return notches[std::size(notches) - 1].second;
+	}
+	static u8 dct03_brake(u8 value)
+	{
+		// Depending on the game, this device presents in either Non Self-Lapping or Self-Lapping mode.
+		// NSL Release   Maintain  Increase  Emergency
+		//     0x23-0x64 0x65-0x89 0x8A-0xD6 0xD7
+		// SL  Released  B1        B2        B3        B4        B5        B6        EB
+		//     0x23-0x2A 0x2B-0x3C 0x3D-0x4E 0x4F-0x63 0x64-0x8A 0x8B-0xB0 0xB1-0xD6 0xD7
+		if (0x18 >= value)
+			return 0x23;
+		if (value >= 0xF8)
+			return 0xD7;
+		// We've trimmed 0x20 (0x8 for EB, 0x18 for release) leaving us with ~0xE0.
+		// 0xD7-0x23=0xB3 represents about 80% of the number space of 0xE0, so compress the remaining input values into that range.
+		u8 offset = 0x9 + value / 85;
+		return value / 5 * 4 + offset;
+	}
+
+#define get_ab(buttons) (button_at(buttons, CID_MC_A) | button_at(buttons, CID_MC_B))
+#define get_cd(buttons) (button_at(buttons, CID_MC_C) | button_at(buttons, CID_MC_D))
+#define swap_cd(buttons) ((button_at(buttons, CID_MC_C) << 1) | (button_at(buttons, CID_MC_D) >> 1))
+#define get_ss(buttons) (button_at(buttons, CID_MC_START) | button_at(buttons, CID_MC_SELECT))
+
+	// MasconControlID buttons are laid out in Type 2 ordering, no need to remap.
+	constexpr u8 dct01_buttons(u8 buttons) { return buttons; }
+	constexpr u8 dct02_buttons(u8 buttons)
+	{
+		return ((get_ab(buttons) << 2) | (swap_cd(buttons) >> 2) | get_ss(buttons));
+	}
+	constexpr u8 dct03_buttons(u8 buttons)
+	{
+		return (get_ab(buttons) | (button_at(buttons, CID_MC_CAMERA) >> 4) | ((get_cd(buttons) | get_ss(buttons)) << 1));
+	}
+
+	int MasconState::TokenIn(u8* buf, int len)
```

Comment:
These TokenIn methods are left over from the time when there were multiple variants, one for each API, now the body can be moved directly in mascon_handle_data()